packages/telemetry/src/flight-recorder.js

```diff
@@ -217,18 +181,186 @@ export const makeMemoryMappedCircularBuffer = async ({
       );
     }
 
-    arena.set(record.subarray(0, firstWriteLength), Number(circEnd));
-    if (firstWriteLength < record.byteLength) {
-      // Write to the beginning of the arena.
-      arena.set(record.subarray(firstWriteLength, record.byteLength), 0);
-    }
+    writeRecord(record, firstWriteLength, circEnd);
+
     header.setBigUint64(
       I_CIRC_END,
       (circEnd + BigInt(record.byteLength)) % arenaSize,
     );
   };
 
   return { readCircBuf, writeCircBuf };
+}
+
+// TODO use .write() instead of .writeSync() because the writes are already
+// buffered earlier in the slog sender infra, and fsync after every block
+/**
+ * Variant of makeMemoryMappedCircularBuffer that writes to a file instead of using BufferFromFile
+ *
+ * @param {{
+ *   circularBufferSize?: number,
+ *   stateDir?: string,
+ *   circularBufferFilename?: string
+ * }} opts
+ * @returns {Promise<CircularBuffer>}
+ */
+export const makeSimpleCircularBuffer = async ({
+  circularBufferSize = DEFAULT_CBUF_SIZE,
+  stateDir = '/tmp',
+  circularBufferFilename,
+}) => {
+  const filename = circularBufferFilename || `${stateDir}/${DEFAULT_CBUF_FILE}`;
+
+  const newArenaSize = await initializeCircularBuffer(
+    filename,
+    circularBufferSize,
+  );
+
+  const file = await fsp.open(filename, 'r+');
+
+  const headerBuffer = Buffer.alloc(I_ARENA_START);
+
+  await file.read({
+    buffer: headerBuffer,
+    length: I_ARENA_START,
+    position: 0,
+  });
+  const header = new DataView(headerBuffer.buffer);
+
+  // Detect the arena size from the header, if not initialized.
+  const hdrArenaSize = header.getBigUint64(I_ARENA_SIZE);
+  const arenaSize = newArenaSize || hdrArenaSize;
+
+  const hdrMagic = header.getBigUint64(I_MAGIC);
+  SLOG_MAGIC === hdrMagic ||
+    Fail`${filename} is not a slog buffer; wanted magic ${SLOG_MAGIC}, got ${hdrMagic}`;
+  arenaSize === hdrArenaSize ||
+    Fail`${filename} arena size mismatch; wanted ${arenaSize}, got ${hdrArenaSize}`;
+
+  /** @type {(outbuf: Uint8Array, readStart: number, firstReadLength: number) => void} */
+  const readRecord = (outbuf, readStart, firstReadLength) => {
+    fs.readSync(file.fd, outbuf, {
+      length: firstReadLength,
+      position: Number(readStart) + I_ARENA_START,
+    });
+    if (firstReadLength < outbuf.byteLength) {
+      fs.readSync(file.fd, outbuf, {
+        offset: firstReadLength,
+        length: outbuf.byteLength - firstReadLength,
+        position: I_ARENA_START,
+      });
+    }
+  };
+
+  /**
+   * Writes to the file, offset by the header size
+   *
+   * @param {Uint8Array} record
+   * @param {number} firstWriteLength
+   * @param {bigint} circEnd
+   */
+  const writeRecord = (record, firstWriteLength, circEnd) => {
+    fs.writeSync(
+      file.fd,
+      record.subarray(0, firstWriteLength),
+      // TS saying options bag not available
+      undefined,
+      undefined,
```

Comment:
I believe it's more efficient to use offset + length

Reply:
You mean instead of `subarray`? makes sense. incoming.